File: App/Resources/AlgorithmDetails/bottomupheapsort/bottomupheapsort.swift
func sort(_ arr: inout [Int]) {
    let n = arr.count

    func siftDown(_ i: Int, _ b: Int) {
        var j = i
        while 2 * j + 1 < b {
            if 2 * j + 2 < b {
                j = arr[2 * j + 2] > arr[2 * j + 1] ? 2 * j + 2: 2 * j + 1
            } else {
                j = 2 * j + 1
            }
        }
        while arr[i] > arr[j] {
            j = (j - 1) / 2
        }
        while j > i {
            arr.swapAt(i, j)
            j = (j - 1) / 2
        }
    }

    for i in stride(from: (n - 1) / 2, through: 0, by: -1) {
        siftDown(i, n)
    }

    for i in stride(from: n - 1, through: 1, by: -1) {
        arr.swapAt(0, i)
        siftDown(0, i)
    }
}

var array: [Int] = [
    0, 39, 21, 62, 91, 77, 14, 23,
    90, 69, 51, 81, 68, 83, 32, 56,
]
sort(&array)
print(array)

File: App/Resources/AlgorithmDetails/bottomupheapsort/bottomupheapsort.cs
using System;

public class BottomUpHeapSort
{
  public static void SiftDown(int[] arr, int i, int b)
  {
    int j = i;
    while (2 * j + 1 < b)
    {
      if (2 * j + 2 < b)
      {
        j = (arr[2 * j + 2] > arr[2 * j + 1]) ? 2 * j + 2 : 2 * j + 1;
      }
      else
      {
        j = 2 * j + 1;
      }
    }
    while (arr[i] > arr[j])
    {
      j = (j - 1) / 2;
    }
    while (j > i)
    {
      (arr[i], arr[j]) = (arr[j], arr[i]);
      j = (j - 1) / 2;
    }
  }

  public static void Sort(int[] arr)
  {
    int n = arr.Length;
    for (int i = (n - 1) / 2; i >= 0; i--)
    {
      SiftDown(arr, i, n);
    }
    for (int i = n - 1; i > 0; i--)
    {
      (arr[0], arr[i]) = (arr[i], arr[0]);
      SiftDown(arr, 0, i);
    }
  }

  public static void Main(String[] args)
  {
    int[] array = { 0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56 };
    Sort(array);
    string result = "[" + String.Join(", ", array) + "]";
    Console.WriteLine(result);
  }
}